### PR TITLE
Add CUDA implementation of kthvalue using radixSelect

### PR DESCRIPTION
**Pull request by pararthshah**
This adds an implementation of THCTensor_(kthvalue) which uses the parallel radixSelect implementation used in topk.

Updated test/test_cuda.py to not skip tests for kthvalue.

### CODE REVIEW

File: aten/src/THC/THCTensorKthValue.cu [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+#include "THCTensorCopy.h"
+#include "THCTensorMath.h"
+#include "THCAsmUtils.cuh"
+#include "THCScanUtils.cuh"
+#include "THCTensorTypeUtils.cuh"
+#include "THCTensorMathReduce.cuh"
+#include <algorithm> // for std::min
+
+#if CUDA_VERSION >= 7000
+#include <thrust/system/cuda/execution_policy.h>
```
Comment (zou3519) [inline aten/src/THC/THCTensorKthValue.cu]: What is this used for?

File: aten/src/THC/THCTensorKthValue.cuh [hunk 2]
```diff
@@ -0,0 +1,67 @@
+#ifndef THC_TENSOR_KTHVALUE_CUH
+#define THC_TENSOR_KTHVALUE_CUH
+
+# include "THCTensorTopK.cuh"
```
Comment (zou3519) [inline aten/src/THC/THCTensorKthValue.cuh]: nit: space between # and include

File: aten/src/THC/generic/THCTensorKthValue.cu [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+  THArgCheck(dim >= 0 && dim < numDims, 6, "dim not in range");
+
+  int64_t sliceSize = THCTensor_(size)(state, input, dim);
+  THArgCheck(k > 0 && k <= sliceSize, 5, "k not in range for dimension");
+
+  // Build the output size, which is the dim being selected set to
+  // size 1
+  THLongStorage* kthValueSize = THCTensor_(newSizeOf)(state, input);
+  THLongStorage_set(kthValueSize, dim, 1);
+  THCTensor_(resize)(state, kthValue, kthValueSize, NULL);
```
Comment (zou3519) [inline aten/src/THC/generic/THCTensorKthValue.cu]: Let's say keepdim is False. Then we should attempt to avoid the resize, because the resize changes the contiguity of the inputs, which could be bad. [kthvalue](https://pytorch.org/docs/master/torch.html?highlight=kth%20value#torch.kthvalue) supports passing in "output" tensors that the results should be stored in.

[This function](https://github.com/pytorch/pytorch/blob/b002aee0ff5eeea7172e4037b3d5936202cb6aef/aten/src/THC/THCTensor.cpp#L367-L381) might help.

File: aten/src/THC/generic/THCTensorKthValue.cu [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    /* This is used as a template parameter to calculate indices. */      \
+    /* We only specialize it if all collapsed dim sizes are the */        \
+    /* same; otherwise, we use -1 which is the specialization */          \
+    /* parameter for arbitrary dimensions */                              \
+    int allDims = inputInfo.dims;                                         \
+    if (kthValueInfo.dims != allDims || indicesInfo.dims != allDims) {    \
+      allDims = -1;                                                       \
+    }                                                                     \
+                                                                          \
+    RUN_DIM(INDEX_T);
```
Comment (zou3519) [inline aten/src/THC/generic/THCTensorKthValue.cu]: nit: unindent one level to match the code above and not the macro indentation

File: aten/src/THC/generic/THCTensorKthValue.cu [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+    // Based on required index size, run the algorithm with the
+    // appropriate index type
+    if (THCTensor_canUse32BitIndexMath(state, input) &&
+        THCTensor_canUse32BitIndexMath(state, kthValue) &&
+        THCTensor_canUse32BitIndexMath(state, indices)) {
+      RUN_T(uint32_t);
```
Comment (zou3519) [inline aten/src/THC/generic/THCTensorKthValue.cu]: Do these have to be macros? It looks like you should be able to template-ize this.

File: aten/src/THC/generic/THCTensorKthValue.cu [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    } else {
+      RUN_T(uint64_t);
```
Comment (zou3519) [inline aten/src/THC/generic/THCTensorKthValue.cu]: Could you try running a test case with large indices? You don't have to add it as a unit test (it would probably cause a lot of GPUs to oom...)

File: aten/src/THC/generic/THCTensorKthValue.cu [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+  #define RUN_T(INDEX_T)                                                  \
+    TensorInfo<real, INDEX_T> inputInfo =                                 \
+      getTensorInfo<real, THCTensor, INDEX_T>(state, input);              \
+    TensorInfo<real, INDEX_T> kthValueInfo =                              \
+      getTensorInfo<real, THCTensor, INDEX_T>(state, kthValue);           \
+    TensorInfo<int64_t, INDEX_T> indicesInfo =                            \
+      getTensorInfo<int64_t, THCudaLongTensor, INDEX_T>(state, indices);  \
+                                                                          \
+    /* We use these structures solely to find the offset to */            \
+    /* each slice we are operating on */                                  \
```
Comment (zou3519) [inline aten/src/THC/generic/THCTensorKthValue.cu]: I'm not sure what's going on here. Why are you setting the sizes of the dimension to 1 and then flattening the tensor? 

Also, shouldn't
`kthValueInfo.sizes[dim]` and  `indicesInfo.sizes[dim]` already be 1 because you resized them?

File: aten/src/THC/generic/THCTensorKthValue.cu [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    inputInfo.sizes[dim] = 1;                                             \
+    kthValueInfo.sizes[dim] = 1;                                          \
+    indicesInfo.sizes[dim] = 1;                                           \
+                                                                          \
+    /* Collapse all other dims */                                         \
+    int collapseInputDim = inputInfo.collapseDims(dim);                   \
+    int collapseKthValueDim = kthValueInfo.collapseDims(dim);             \
+    int collapseIndicesDim = indicesInfo.collapseDims(dim);               \
+                                                                          \
+    int64_t inputSlices = 1;                                              \
```
Comment (zou3519) [inline aten/src/THC/generic/THCTensorKthValue.cu]: What exactly is a "inputSlice"?

**Review by ssnl — COMMENTED** [review]
Does this backprop properly?

File: aten/src/THC/THCTensorKthValue.cu (hunk 1 above)
Comment (pararthshah) [reply aten/src/THC/THCTensorKthValue.cu]: This was unnecessary. Removed it.

File: aten/src/THC/THCTensorKthValue.cuh (hunk 2 above)
Comment (pararthshah) [reply aten/src/THC/THCTensorKthValue.cuh]: done

File: aten/src/THC/generic/THCTensorKthValue.cu (hunk 4 above)
Comment (pararthshah) [reply aten/src/THC/generic/THCTensorKthValue.cu]: done

File: aten/src/THC/generic/THCTensorKthValue.cu (hunk 5 above)
Comment (pararthshah) [reply aten/src/THC/generic/THCTensorKthValue.cu]: I've followed the style in THCTensorTopK.cu, which defines these as macros. Are templates preferable to macros?

File: aten/src/THC/generic/THCTensorKthValue.cu (hunk 8 above)
Comment (pararthshah) [reply aten/src/THC/generic/THCTensorKthValue.cu]: Suppose the input tensor shape is (M1, M2, M3, M4) and dim=2. We need to select the kth largest elements in the M3 dimension of every slice of the tensor, so there will be a total of M1xM2xM4 slices along which selection is to be done. We are parallelizing this across blocks by spawning a separate block for each slice along which we need to select, so number of blocks = inputSlices = M1xM2xM4.

This is the same logic as used in THCTensorTopK.cu, so I assumed it was self-explanatory. But I'm happy to add documentation to both files.

File: aten/src/THC/generic/THCTensorKthValue.cu (hunk 3 above)
Comment (pararthshah) [reply aten/src/THC/generic/THCTensorKthValue.cu]: Would the same issue occur in THCTensorTopK, which does the resize as well? https://github.com/pytorch/pytorch/blob/master/aten/src/THC/generic/THCTensorTopK.cu#L27

Maybe this should be done in a follow-up PR to fix both implementations.

Comment (zou3519) [reply aten/src/THC/generic/THCTensorKthValue.cu]: True, let's leave this then.

File: aten/src/THC/generic/THCTensorKthValue.cu (hunk 8 above)
Comment (zou3519) [reply aten/src/THC/generic/THCTensorKthValue.cu]: Understood. Documentation would be good; I have a hard time reading THCTensorTopK.cu

**Review by fmassa — COMMENTED** [review]
I don't think there is much left for this PR to get merged.

**Comment by t-vi** [comment]
@pararthshah anything I can be of help with? I'd really love to see KthValue in Cuda.

**Comment by fmassa** [comment]
@t-vi I think @pararthshah is not working on this anymore.
And indeed, it would be awesome to have kthvalue on CUDA.

@t-vi do you think you would have the chance to look into finishing it?

**Comment by pararthshah** [comment]
@t-vi It would be great if you can finish this. Thank you!

**Comment by t-vi** [comment]
OK, thanks for the quick reply!

**Comment by SsnL** [comment]
merged as part of #17544
